internal/provider/backuppolicy/resource.go

```diff
@@ -0,0 +1,119 @@
+// (C) Copyright 2013-2024 Dassault Systemes SE.  All Rights Reserved.
+//
+// This software is licensed under a BSD 3-Clause License.
+// See the LICENSE file provided with this software.
+
+package backuppolicy
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/nuodb/terraform-provider-nuodbaas/internal/framework"
+	"github.com/nuodb/terraform-provider-nuodbaas/internal/helper"
+	"github.com/nuodb/terraform-provider-nuodbaas/openapi"
+
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+)
+
+var (
+	_ framework.ResourceState   = &BackupPolicyResourceModel{}
+	_ framework.DataSourceState = &BackupPolicyResourceModel{}
+)
+
+type BackupPolicyResourceModel openapi.BackupPolicyModel
+
+func (state *BackupPolicyResourceModel) Reset() {
+	*state = BackupPolicyResourceModel{}
+}
+
+func (state *BackupPolicyResourceModel) CheckReady(ctx context.Context, client openapi.ClientInterface) error {
+	return nil
+}
+
+func (state *BackupPolicyResourceModel) Create(ctx context.Context, client openapi.ClientInterface) error {
+	resp, err := client.CreateBackupPolicy(ctx, state.Organization, state.Name, openapi.BackupPolicyModel(*state))
+	if err != nil {
+		return err
+	}
+	return helper.ParseResponse(resp, nil)
+}
+
+func (state *BackupPolicyResourceModel) Read(ctx context.Context, client openapi.ClientInterface) error {
+	resp, err := client.GetBackupPolicy(ctx, state.Organization, state.Name)
+	if err != nil {
+		return err
+	}
+	state.Reset()
+	return helper.ParseResponse(resp, state)
+}
+
+func (state *BackupPolicyResourceModel) Update(ctx context.Context, client openapi.ClientInterface, currentState framework.ResourceState) error {
+	// Fetch backup policy and get resourceVersion
+	latest := &BackupPolicyResourceModel{
+		Organization: state.Organization,
+		Name:         state.Name,
+	}
+	err := latest.Read(ctx, client)
+	if err != nil {
+		return err
+	}
+	for {
+		state.ResourceVersion = latest.ResourceVersion
+		resp, err := client.CreateBackupPolicy(ctx, state.Organization, state.Name, openapi.BackupPolicyModel(*state))
+		if err != nil {
+			return err
+		}
+		// Decode the response and check that there is no error
+		err = helper.ParseResponse(resp, nil)
+		if err == nil {
+			return nil
+		}
+		// If error is not retriable (code=CONCURRENT_UPDATE), fail fast
+		if apiError, ok := err.(*helper.ApiError); !ok || apiError.GetCode() != openapi.CONCURRENTUPDATE {
+			return err
+		}
+		// Re-fetch policy and get resourceVersion
+		err = latest.Read(ctx, client)
+		if err != nil {
+			return err
+		}
+	}
+}
```

Comment:
Most of this code seems to be the same for all resources. Is it possible to factor out the duplicate code?

Reply:
I remember looking into it in the past and deciding that it wasn't worth it because the database update required special processing for the DBA password anyway. Maybe it is worth it now that we have other resources that duplicate what the project resource does.